pkg/kv/kvserver/replica_raft.go

```diff
@@ -1141,6 +1172,30 @@ func (r *Replica) refreshProposalsLocked(
 		}
 	}
 
+	r.store.metrics.SlowRaftRequests.Update(slowProposalCount)
+
+	// If the breaker isn't tripped yet but we've detected commands that have
+	// taken too long to replicate, trip the breaker now.
+	//
+	// NB: we still keep reproposing commands on this and subsequent ticks
+	// even though this seems strictly counter-productive, except perhaps
+	// for the probe's proposals. We could consider being more strict here
+	// which could avoid build-up of raft log entries during outages, see
+	// for example:
+	// https://github.com/cockroachdb/cockroach/issues/60612
+	if r.breaker.Signal().Err() == nil && maxInflightDuration > 0 {
+		log.Warningf(ctx,
+			"have been waiting %.2fs for slow proposal %s",
+			maxInflightDuration.Seconds(), maxInflightDurationReq,
+		)
+		// NB: this is async because we're holding lots of locks here, and we want
+		// to avoid having to pass all the information about the replica into the
+		// breaker (since the breaker needs access to this information at will to
+		// power the probe anyway). Over time, we anticipate there being multiple
+		// mechanisms which trip the breaker.
+		r.breaker.TripAsync()
```

Comment:
Since this is async, could we end up logging the above multiple times if we're called again before the breaker actually trips?

Reply:
Yeah, 3s seems reasonably unlikely.